File: src/server/components/Tool.ts
/* eslint-disable roblox-ts/lua-truthiness */
import { OnInit, OnStart } from "@flamework/core";
import { Component, BaseComponent } from "@flamework/components";
import Signal from "@rbxts/signal";
import { HttpService } from "@rbxts/services";
import { Events } from "server/events";
import { Janitor } from "@rbxts/janitor";
import { Players } from "@rbxts/services";
import { ParseInput } from "server/modules/InputParser";
import { Action } from "server/modules/Action";
import { Defer } from "server/modules/Defer";
import { TypePredicate } from "typescript";
import { Tools } from "shared/Config";
import { deepCopy } from "@rbxts/object-utils";

export type ITool = Tool<ToolAttributes, ToolInstance>;

export type ActionInfo = {
	Action: string;
	Mobile?: {};
};

export type InputInfo = {
	// "EQUIPPED"
	[toolState: string]: {
		// "Begin"
		[UserInputState: string]: {
			// "One", "W", "P", ...
			[UserInput: string]: ActionInfo;
		};
	};
};

export interface ToolInstance extends Model {
	BodyAttach: BasePart;
}

export interface ToolAttributes {
	BUTTON_TOGGLE: string;
}

export interface Actions {
	[actionName: string]: Action;
}

type ToolSignal = <T extends Tool<ToolAttributes, ToolInstance>>(tool: T) => void;
const ToolAdded = new Signal<ToolSignal>();
export { ToolAdded };

@Component()
export abstract class Tool<A extends ToolAttributes, I extends ToolInstance>
	extends BaseComponent<A, I>
	implements OnStart
{
	public abstract Incompatible: string[];
	public abstract Actions: Actions;
	public abstract InputInfo: InputInfo;
	public abstract Init?(): void;
	public abstract PlayerInit?(): void;

	protected janitor = new Janitor();
	private ButtonedInputInfo?: InputInfo;
	public id = HttpService.GenerateGUID();
	public state = "nil";
	public timeCreated = tick();
	public Player?: Player;
	public stateChanged = new Signal<(state: string) => void>();

	public getState() {
		return this.state;
	}

	public setState(state: string) {
		this.state = state;
		this.stateChanged.Fire(state);
	}

	onStart() {
		// Not sure if using promise.defer is a hacky way or a legit way to acces the abstract methods and properties. Trying to access in onStart() without deferring will result in a nil value.
		Defer(() => {
			this.maid.GiveTask(this.janitor);

			this.ManageButtons();
			this.ManageAncestry();

			ToolAdded.Fire(this);

			if (this.Init !== undefined) {
				this.Init();
			}
		});
	}

	private ManageButtons() {
		for (const [key, value] of pairs(this.attributes as unknown as Map<never, never>)) {
			if (this.isButton(key)) {
				this.onAttributeChanged(key, (name) => {
					this.UpdateButtons();
				});
			}
		}
		this.UpdateButtons();
	}

	private ManageAncestry() {
		this.UpdateAncestry();
		this.instance.AncestryChanged.Connect(() => this.UpdateAncestry());
	}

	private UpdateAncestry() {
		if (!this.instance.IsDescendantOf(game)) {
			pcall(() => {
				this.instance.Destroy();
			});
		} else if (Players.GetPlayerFromCharacter(this.instance.Parent) !== this.Player) {
			this.janitor.Cleanup();
			this.InitPlayer();
		} else {
			this.janitor.Cleanup();
			this.InitWorkspace();
		}
	}

	private RequirePlayer() {
		const Player = Players.GetPlayerFromCharacter(this.instance.Parent);
		if (!Player) {
			error("attempting to call a function which requires the tool to be under a player");
		}
		return Player;
	}

	private InitPlayer() {
		this.Player = this.RequirePlayer();
		this.janitor.Add(this.SetupInput());
		this.Input("INIT", { Input: "None", State: "None" });

		if (this.PlayerInit !== undefined) {
			this.PlayerInit();
		}
	}

	private InitWorkspace() {
		warn("this section // for dropping items // has not yet been implemented, did you mean to do this?");
	}

	private SetupInput() {
		this.RequirePlayer();
		return Events.Input.connect((player, inputobject) => {
			if (this.Player === player) {
				const Input = ParseInput(inputobject);
				this.Input(this.getState(), Input);
			}
		});
	}

	private Input(state: string, input: { Input: string; State: string }) {
		this.RequirePlayer();
		const ActionInfo = this.GetActionInfo(state, input);
		if (ActionInfo === undefined) {
			return;
		}
		const Action = this.Actions[ActionInfo.Action];
		if (!Action) {
			error(`action ${ActionInfo.Action} does not exist`);
		}

		this.AddAction(Action);
	}

	private AddAction(Action: Action) {
		Action.Start();
	}

	private GetActionInfo(toolState: string, input: { Input: string; State: string }): ActionInfo | undefined {
		const InputInfo = this.ButtonedInputInfo;
		if (!InputInfo) {
			error();
		}
		if (
			!InputInfo[toolState] ||
			!InputInfo[toolState][input.State] ||
			!InputInfo[toolState][input.State][input.Input]
		) {
			return;
		}

		return InputInfo[toolState][input.State][input.Input];
	}

	/** updates the ButtonedInputInfo to be accurate to newly assigned buttons */
	private UpdateButtons() {
		this.ButtonedInputInfo = this.fillButtonsInTable(this.InputInfo);
	}

	/**
	 *
	 * Takes in a table and loops through its indexes and values. Ensures that all of it's indexes if buttons are set to their actual attribute value and if any of it's values are tables that their indexes are also checked and made sure that any buttons in their indexes are set to their actual attribute values aswell.
	 *
	 * @example
	 *
	 * ```typescript
	 * const Table: InputInfo = {
	 * 		Holstered: {
	 *			Begin: {
	 *				BUTTON_TOGGLE: {
	 *					Action = "Equip"
	 *				}
	 *			}
	 * 		}
	 * }
	 *
	 * const ButtonedInputInfo = fillButtonsInTable(Table) where this.attributes = { BUTTON_TOGGLE: "One" } // "One" being representative of the 1 key on the keyboard
	 *
	 * // what ButtonedInputInfo is equal to:
	 * // notice the purpose is to remove all BUTTON_ and replace them with their actual value
	 * {
	 * 		Holstered: {
	 *			Begin: {
	 *				One: {
	 *					Action = "Equip"
	 *				}
	 *			}
	 *		}
	 * 	}
	 * ```
	 *
	 * @param Table a table which has a string index
	 * @returns A Table who's indexes if buttons were replaced by their actual value
	 */
	private fillButtonsInTable<T extends { [key: string]: unknown }>(Table: T) {
		Table = deepCopy(Table);
		const newTable = {};

		for (let [key, value] of pairs(Table as unknown as Map<never, never>)) {
			// EXAMPLE: key might equal "CUSTOM_TOGGLE" and this.attributes might equal {"CUSTOM_TOGGLE": "MouseButton1"}
			if (this.isButton(key)) {
				const Button = this.instance.GetAttribute(key);
				if (Button && typeIs(Button, "string")) {
					key = Button as never;
				}
			}

			if (typeIs(value, "table")) {
				value = this.fillButtonsInTable(value);
			}

			newTable[key] = value;
		}

		return newTable as T;
	}

	/** returns if the string is an identifier for a button */
	private isButton(str: string): boolean {
		return string.sub(str, 0, 7) === "BUTTON_";
	}

	protected GetCharPlayer() {
		const Player = this.Player;
		if (!Player) {
			error("player required");
		}
		const Character = Player.Character;
		if (!Character) {
			error("character required");
		}

		return [Player, Character] as [Player, Model];
	}
}

File: src/server/Components/Tool.lua
local Knit = require(game:GetService("ReplicatedStorage").Knit)
local Network = require(game:GetService("ReplicatedStorage").Network)
local HttpService = game:GetService("HttpService")
local Config = require(game:GetService("ReplicatedStorage").Config)
local Janitor = require(Knit.Util.Janitor)
local Component = require(Knit.Util.Component)
local Signal = require(Knit.Util.Signal)
local RemoteSignal = require(Knit.Util.Remote.RemoteSignal)

local Players = game:GetService("Players")

local ActionHandlers = script.Parent.Parent.Modules.Action.ActionHandler

local Tools = {}

local Tool = {}
Tool.__index = Tool
Tool.Tag = "Tool"

Tool.CameraDirection = Signal.new()
Tool.CameraDirections = {}

Network:BindEvents({
    CameraDirection = function(player: Player, direction: string)
        assert(typeof(direction) == "string" and (direction == "Up" or direction == "Down"))
        Tool.CameraDirections[player] = direction
    end
})

function Tool.new(instance)

    local self = setmetatable({}, Tool)

    self._janitor = Janitor.new()
    self._janitor:Add(instance)

    self.Config = Config.Tools[instance.Name]
	if not self.Config then error("tool " .. instance.Name .. " need to be configured") end

    self.Name = self.Config.Name
    self.ActionHandler = require(ActionHandlers:FindFirstChild(self.Config.ActionHandler))
	self.DefaultHandler = self.ActionHandler
	
    self.Id = HttpService:GenerateGUID()
    self.Buttons = {}

    Tools[instance] = self
    Tools[self.Id] = self

    return self

end

function Tool:Init()
    local player = Players:GetPlayerFromCharacter(self.Instance.Parent)

    if player then self:CharacterSetup(player)
    elseif Config.Tools.Droppable then self:DroppableSetup() 
    else self:Destroy() end
end

function Tool:CharacterSetup(player: Player) -- setup tool for usage as a child of a character
    self.Mode = "Player"

    self.PlayerJanitor = Janitor.new() 
    self._janitor:Add(self.PlayerJanitor)

    self.StateChanged = Signal.new(self.PlayerJanitor)
	self.HandlerChanged = Signal.new(self.PlayerJanitor)

    self.Player = player
    self.Character = self.Instance.Parent

    self:ManageSiblings()
    self:ManageInputs()

    self.Actions = {}
    self:Input(Enum.UserInputState.None,Enum.UserInputState.None)
end

function Tool:ManageInputs()
    self.InputEnabled = true
    if not Tools[self.Player] then
        Tools[self.Player] = {}
    end
    table.insert(Tools[self.Player], self)
end


function Tool:Lock(signal, actionHandler)
	if self.Locked then error("cannot lock when already locked") end
	self.Locked = true
	self.ActionHandler = actionHandler
	
	signal:Connect(function() 
		self.Locked = nil
        self:ChangeHandler(self.DefaultHandler)
	end)
	
end

function Tool:Damage(humanoid,damage)
    humanoid:TakeDamage(damage)
    local damagedPlayer = Players:GetPlayerFromCharacter(humanoid.Parent)
    if damagedPlayer then Network:FireClient(damagedPlayer, "Damaged") end
    Network:FireClient(self.Player, "Hit")
end

function Tool:ManageSiblings()
    self.Siblings = {}
	
	local function TryAddSibling(child)
		if Tools[child] then
			self.Siblings[child] = Tools[child]
		end
	end

    self.PlayerJanitor:Add(self.Character.ChildAdded:Connect(TryAddSibling))
	self.PlayerJanitor:Add(self.Character.ChildRemoved:Connect(function(child)
        if not self.Siblings[child] then return end  
		self.Siblings[child] = nil
    end))
	
	for _,child in pairs(self.Character:GetChildren()) do TryAddSibling(child) end
end

function Tool:CharacterDestroy()
    table.remove(Tools[self.Player], table.find(Tools[self.Player], self))
    self.Mode = nil
    self.PlayerJanitor:Destroy()
end

function Tool.handleInput() -- DNT (do not trust) : recieved by client
    Network:BindEvents({
        Input = function(player, inputState, inputObject)
            if Tools[player] then
                assert(typeof(inputObject) == "EnumItem" and typeof(inputState) == "EnumItem")
                for _, tool in pairs(Tools[player]) do
                    tool:Input(inputState,inputObject)
                end
            end
        end
    })
end

function Tool:ChangeHandler(handler)
	self.ActionHandler = handler
	self.HandlerChanged:Fire()
end

function Tool:Input(inputState : EnumItem?, inputObject: EnumItem?)
    
    local Action = self.ActionHandler:GetAction(self.State, inputState, inputObject)
    if Action then self:AddAction(Action) end
end


function Tool:AddAction(Action)
    Action:SetPrimaryTool(self)
    self.Actions[Action.Name] = Action
    self.PlayerJanitor:Add(Action.Ended:Connect(function()
        self.Actions[Action.Name] = nil
    end))
    Action:Start()
end

function Tool:Queue(Action, ActionCaller)
	self.PlayerJanitor:Add(ActionCaller.Ended:Connect(function()
		self:AddAction(Action)	
	end))
end

function Tool:ChangeState(state)
    self.State = state
    self.StateChanged:Fire()
end


function Tool:Destroy()
    if self.Mode == "Player" then
        self:CharacterDestroy()
    end
    Tools[self.Instance] = nil
    Tools[self.Id] = nil
    self._janitor:Destroy()
end


Tool.handleInput()

return Tool
